File: Stripe/STPCardParams.m
//
//  STPCardParams.m
//  Stripe
//
//  Created by Jack Flintermann on 10/4/15.
//  Copyright © 2015 Stripe, Inc. All rights reserved.
//

#import "STPCardParams.h"
#import "STPCard+Private.h"

#import "STPCardValidator.h"
#import "StripeError.h"

@implementation STPCardParams

@synthesize additionalAPIParameters = _additionalAPIParameters;

- (instancetype)init {
    self = [super init];
    if (self) {
        _additionalAPIParameters = @{};
        _address = [STPAddress new];
    }
    return self;
}

- (NSString *)last4 {
    if (self.number && self.number.length >= 4) {
        return [self.number substringFromIndex:(self.number.length - 4)];
    } else {
        return nil;
    }
}

- (void)setName:(NSString *)name {
    _name = [name copy];
    self.address.name = self.name;
}

- (void)setAddress:(STPAddress *)address {
    _address = address;
    _name = [address.name copy];
}

#pragma mark - Description

- (NSString *)description {
    NSArray *props = @[
                       // Object
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],

                       // Basic card details
                       [NSString stringWithFormat:@"last4 = %@", self.last4],
                       [NSString stringWithFormat:@"expMonth = %lu", (unsigned long)self.expMonth],
                       [NSString stringWithFormat:@"expYear = %lu", (unsigned long)self.expYear],
                       [NSString stringWithFormat:@"cvc = %@", (self.cvc) ? @"<redacted>" : nil],

                       // Additional card details (alphabetical)
                       [NSString stringWithFormat:@"currency = %@", self.currency],

                       // Cardholder details
                       [NSString stringWithFormat:@"name = %@", (self.name) ? @"<redacted>" : nil],
                       [NSString stringWithFormat:@"address = %@", (self.address) ? @"<redacted>" : nil],
                       ];

    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPFormEncodable

+ (NSString *)rootObjectName {
    return @"card";
}

+ (NSDictionary *)propertyNamesToFormFieldNamesMapping {
    return @{
             NSStringFromSelector(@selector(number)): @"number",
             NSStringFromSelector(@selector(cvc)): @"cvc",
             NSStringFromSelector(@selector(name)): @"name",
             NSStringFromSelector(@selector(addressLine1)): @"address_line1",
             NSStringFromSelector(@selector(addressLine2)): @"address_line2",
             NSStringFromSelector(@selector(addressCity)): @"address_city",
             NSStringFromSelector(@selector(addressState)): @"address_state",
             NSStringFromSelector(@selector(addressZip)): @"address_zip",
             NSStringFromSelector(@selector(addressCountry)): @"address_country",
             NSStringFromSelector(@selector(expMonth)): @"exp_month",
             NSStringFromSelector(@selector(expYear)): @"exp_year",
             NSStringFromSelector(@selector(currency)): @"currency",
             };
}

#pragma mark - Deprecated methods

- (void)setAddressLine1:(NSString *)addressLine1 {
    self.address.line1 = addressLine1;
}

- (NSString *)addressLine1 {
    return self.address.line1;
}

- (void)setAddressLine2:(NSString *)addressLine2 {
    self.address.line2 = addressLine2;
}

- (NSString *)addressLine2 {
    return self.address.line2;
}

- (void)setAddressZip:(NSString *)addressZip {
    self.address.postalCode = addressZip;
}

- (NSString *)addressZip {
    return self.address.postalCode;
}

- (void)setAddressCity:(NSString *)addressCity {
    self.address.city = addressCity;
}

- (NSString *)addressCity {
    return self.address.city;
}

- (void)setAddressState:(NSString *)addressState {
    self.address.state = addressState;
}

- (NSString *)addressState {
    return self.address.state;
}

- (void)setAddressCountry:(NSString *)addressCountry {
    self.address.country = addressCountry;
}

- (NSString *)addressCountry {
    return self.address.country;
}

#pragma mark - NSCopying

- (id)copyWithZone:(__unused NSZone *)zone {
    STPCardParams *copyCardParams = [self.class new];

    copyCardParams.number = self.number;
    copyCardParams.expMonth = self.expMonth;
    copyCardParams.expYear = self.expYear;
    copyCardParams.cvc = self.cvc;

    // Use ivar to avoid setName:/setAddress: behavior that'd possibly overwrite name/address.name
    copyCardParams->_name = self.name;
    copyCardParams->_address = [self.address copy];

    copyCardParams.currency = self.currency;
    copyCardParams.additionalAPIParameters = self.additionalAPIParameters;

    return copyCardParams;
}

@end

File: Stripe/STPCardParams.swift
//
//  STPCardParams.swift
//  Stripe
//
//  Created by Jack Flintermann on 10/4/15.
//  Copyright © 2015 Stripe, Inc. All rights reserved.
//

import Foundation

/// Representation of a user's credit card details. You can assemble these with
/// information that your user enters and then create Stripe tokens with them using
/// an STPAPIClient.
/// - seealso: https://stripe.com/docs/api#cards
public class STPCardParams: NSObject, STPFormEncodable, NSCopying {
    public var additionalAPIParameters: [AnyHashable: Any] = [:]

    /// The card's number.
    @objc public var number: String?

    /// The last 4 digits of the card's number, if it's been set, otherwise nil.
    @objc
    public func last4() -> String? {
        if number != nil && (number?.count ?? 0) >= 4 {
            return (number as NSString?)?.substring(from: (number?.count ?? 0) - 4)
        } else {
            return nil
        }
    }

    /// The card's expiration month.
    @objc public var expMonth: UInt = 0
    /// The card's expiration year.
    @objc public var expYear: UInt = 0
    /// The card's security code, found on the back.
    @objc public var cvc: String?

    /// The cardholder's name.
    /// @note Changing this property will also changing the name of the
    /// param's `address` property.
    @objc public var name: String? {
        didSet {
            address.name = name
        }
    }

    /// The cardholder's address.
    /// @note Setting `address` to a new value will also change the `name` property to
    /// be the value of `address.name`. However, changing `address.name` directly will
    ///not* change `name`.
    @objc public var address: STPAddress {
        didSet {
            name = address.name
        }
    }
    /// Three-letter ISO currency code representing the currency paid out to the bank
    /// account. This is only applicable when tokenizing debit cards to issue payouts
    /// to managed accounts. You should not set it otherwise. The card can then be
    /// used as a transfer destination for funds in this currency.
    @objc public var currency: String?

    // MARK: - Deprecated methods

    /// The first line of the cardholder's address
    @objc public var addressLine1: String? {
        get {
            return address.line1
        }
        set(addressLine1) {
            address.line1 = addressLine1
        }
    }

    /// The second line of the cardholder's address
    @objc public var addressLine2: String? {
        get {
            return address.line2
        }
        set(addressLine2) {
            address.line2 = addressLine2
        }
    }

    /// The city of the cardholder's address
    @objc public var addressCity: String? {
        get {
            return address.city
        }
        set(addressCity) {
            address.city = addressCity
        }
    }

    /// The state of the cardholder's address
    @objc public var addressState: String? {
        get {
            return address.state
        }
        set(addressState) {
            address.state = addressState
        }
    }

    /// The zip code of the cardholder's address
    @objc public var addressZip: String? {
        get {
            return address.postalCode
        }
        set(addressZip) {
            address.postalCode = addressZip
        }
    }

    /// The country of the cardholder's address
    @objc public var addressCountry: String? {
        get {
            return address.country
        }
        set(addressCountry) {
            address.country = addressCountry
        }
    }

    /// Initializes an empty STPCardParams.
    public override init() {
        address = STPAddress()
        super.init()
        additionalAPIParameters = [:]
    }

    // MARK: - Description
    /// :nodoc:
    @objc public override var description: String {
        let props = [
            // Object
            String(format: "%@: %p", NSStringFromClass(STPCardParams.self), self),
            // Basic card details
            "last4 = \(last4() ?? "")",
            String(format: "expMonth = %lu", UInt(expMonth)),
            String(format: "expYear = %lu", UInt(expYear)),
            "cvc = \(((cvc) != nil ? "<redacted>" : nil) ?? "")",
            // Additional card details (alphabetical)
            "currency = \(currency ?? "")",
            // Cardholder details
            "name = \(((name) != nil ? "<redacted>" : nil) ?? "")",
            "address = <redcated>",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

    // MARK: - STPFormEncodable
    public class func rootObjectName() -> String? {
        return "card"
    }

    public class func propertyNamesToFormFieldNamesMapping() -> [String: String] {
        return [
            NSStringFromSelector(#selector(getter:number)): "number",
            NSStringFromSelector(#selector(getter:cvc)): "cvc",
            NSStringFromSelector(#selector(getter:name)): "name",
            NSStringFromSelector(#selector(getter:addressLine1)): "address_line1",
            NSStringFromSelector(#selector(getter:addressLine2)): "address_line2",
            NSStringFromSelector(#selector(getter:addressCity)): "address_city",
            NSStringFromSelector(#selector(getter:addressState)): "address_state",
            NSStringFromSelector(#selector(getter:addressZip)): "address_zip",
            NSStringFromSelector(#selector(getter:addressCountry)): "address_country",
            NSStringFromSelector(#selector(getter:expMonth)): "exp_month",
            NSStringFromSelector(#selector(getter:expYear)): "exp_year",
            NSStringFromSelector(#selector(getter:currency)): "currency",
        ]
    }

    // MARK: - NSCopying
    /// :nodoc:
    @objc
    public func copy(with zone: NSZone? = nil) -> Any {
        let copyCardParams = STPCardParams()

        copyCardParams.number = number
        copyCardParams.expMonth = expMonth
        copyCardParams.expYear = expYear
        copyCardParams.cvc = cvc

        // Use ivar to avoid setName:/setAddress: behavior that'd possibly overwrite name/address.name
        copyCardParams.name = name
        copyCardParams.address = address.copy() as! STPAddress

        copyCardParams.currency = currency
        copyCardParams.additionalAPIParameters = additionalAPIParameters

        return copyCardParams
    }
}
